### DISCUSSION TITLE
Custom Resource GetAtt not working

### DISCUSSION THREAD

**Discussion by YamiDaisuke** [aws-samples/aws-cdk-examples#641] [Q&A]
### Guidance Question

Custom Resource GetAtt not working

### The Question

Hi!, I'm trying to use a CustomResource to create a MediaTailor channel, so far the scripts go well and the channel gets created but I have two issues:

## First problem:
From my lambda function I'm returning the following object:
```json
{
    "Status": "SUCCESS",
    "PhysicalResourceId": "mediaTailorPlaybackConfigName",
    "StackId": "arn:aws:cloudformation:us-east-1:.....",
    "RequestId": "......",
    "LogicalResourceId": "mediaTailorresource7FE79038",
    "Data": {
        "Arn": "arn:aws:mediatailor:us-east-1:XXXXX:playbackConfiguration/mediaTailorPlaybackConfigName",
        "HLS": "https://xxxxxxxxxxxxxxxxxxx.mediatailor.us-east-1.amazonaws.com/v1/master/xxxxxxxxxxx",
        "DASH": "https://xxxxxxxxxxxxxxxxx.mediatailor.us-east-1.amazonaws.com/v1/dash/xxxxxxxxxxxxx"
    }
}
```

When I try to use `resource.getAtt('HLS').toString()` I get the following error:
```
CustomResource attribute error: Vendor response doesn't contain HLS key in object arn:aws:cloudformation:us-east-1:829011433133:stack/xxxxxxxx|mediaTailorResourcexxxxxxxxxxxxxxx in S3 bucket cloudformation-custom-resource-storage-useast1
```

The second case happens during rollback, even though I'm returning a "PhysicalResourceId" the "DELETE" action receives the same "RequestId" as "PhysicalResourceId" so I don't have a way to find the actual resource and delete it 

### CDK CLI Version

1.102.0

### Example
```js
// For Lambda handler I'm currently using this sample script: 
var response = require('cfn-response');
exports.handler = function (event, context) {
    console.log('REQUEST RECEIVED:\n' + JSON.stringify(event));

    if (event.RequestType === 'Create') {
        console.log('CREATE!');
        response.send(event, context, response.SUCCESS, { 'Message': 'Resource creation successful!', 'HLS': 'hellofromtheotherside.com' });
    } else if (event.RequestType === 'Update') {
        console.log('UDPATE!');
        response.send(event, context, response.SUCCESS, { 'Message': 'Resource update successful!', 'HLS': 'hellofromtheotherside.com' });
    } else if (event.RequestType === 'Delete') {
        console.log('DELETE!');
        response.send(event, context, response.SUCCESS, { 'Message': 'Resource deletion successful!', 'HLS': 'hellofromtheotherside.com' });
    } else {
        console.log('FAILED!');
        response.send(event, context, response.FAILED);
    }
};



// CDK Construct 
class MediaTailorConfig extends cdk.Construct {

    constructor(scope, id, props) {
        super(scope, id, props);

        const onEvent = new lambda.SingletonFunction(this, 'resource', {
            uuid: 'xxxxxxxxxxx',
            code: lambda.Code.fromAsset(path.join(__dirname, 'cfn-mediatailor/index.js.zip'), { encoding: 'utf-8' }),
            handler: 'index.handler',
            timeout: cdk.Duration.seconds(300),
            runtime: lambda.Runtime.NODEJS_14_X,
        });

        const myProvider = new cr.Provider(this, 'MediaTailorProvider', {
            onEventHandler: onEvent,
            logRetention: logs.RetentionDays.TWO_WEEKS, 
        });

        const resource = new cdk.CustomResource(this, 'resource', { serviceToken: myProvider.serviceToken, properties: props });

        this.hls = resource.getAtt('HLS').toString();
    }

}
```

### OS

macOS

### Language

Typescript

**Comment by bkchy** [comment]
We are also having this issue but we are using 2.8.0:

    "aws-cdk": "^2.8.0",
    "aws-cdk-lib": "^2.8.0",

Is there any update on this or has anyone found a workaround?

**Comment by lacteolus** [comment]
The same issue with CDK 2.20.0

**Comment by mlnrt** [comment]
I had the same issue. The way I fixed this is by removing the Provider, and passing the Lambda function ARN directly as the service token.
Essentially doing:

`  class MediaTailorConfig extends cdk.Construct {

        constructor(scope, id, props) {
            super(scope, id, props);

            const onEvent = new lambda.SingletonFunction(this, 'resource', {
                uuid: 'xxxxxxxxxxx',
                code: lambda.Code.fromAsset(path.join(__dirname, 'cfn-mediatailor/index.js.zip'), { encoding: 'utf-8' }),
                handler: 'index.handler',
                timeout: cdk.Duration.seconds(300),
                runtime: lambda.Runtime.NODEJS_14_X,
            });

            const resource = new cdk.CustomResource(this, 'resource', { serviceToken: onEvent.functionArn, properties: props });

            this.hls = resource.getAtt('HLS').toString();
        }
    }`

**Reply by BwL1289** [reply]
This was extremely helpful, thank you @mlnrt. Days of debugging and dozens of threads to find this one comment on GH. Docs for `AwsCustomResource` and `Provider` are lacking in this area.

**Reply by gagipro** [reply]
Hello,

same here, the example is extremely confusing and we don't know if it's a bug or a feature...

Thanks @mlnrt 

Following this one !

**Comment by xiyinzhi** [comment]
I ran into the same issue. The way I fix it is removing `cfn-response` from `async` lambda handler. I believe the root cause is misusing `cfn-response` module. While custom resource provider is listening to this lambda, response directly sent by `cfn-response` will not be received by provider and cause the missing of `data` in provider response. Instead,  we can directly wrap the response and send it back to provider.

Providing an example:

lambda handler:
```
export const handler = async (event: CdkCustomResourceEvent): Promise<CdkCustomResourceResponse> => {
......
    return {
      Data: {
        AccountId: '000011112222',
      },
    };
};
```
Alternatively, if you don't want to use CdkCustomResourceResponse you can simply send the data back:
```
    return {AccountId: '000011112222'};
```


CDK Construct:
```
class abcLambda extends Construct {

    constructor(scope, id, props) {
        super(scope, id, props);

        const onEvent = new SingletonFunction(this, 'resource', {
            uuid: 'xxxxxxxxxxx',
            code: Code.fromAsset(path.join(__dirname, '......'), { encoding: 'utf-8' }),
            handler: 'index.handler',
            timeout: Duration.seconds(300),
            runtime: Runtime.NODEJS_14_X,
        });
    
        const provider = new Provider(this, 'Provider', {
          onEventHandler: onEvent,
        });
    
        const cr = new CustomResource(this, `lambda-${name}`, {
          serviceToken: provider.serviceToken,
          properties: {
            name: name,
            date: new Date().toISOString(),
          },
        });
    
        return cr.getAttString('AccountId');
        }

}
```


ref: 
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-lambda-function-code-cfnresponsemodule.html
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html
